src/hotspot/share/utilities/intn_t.hpp

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright (c) 2025, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef SHARE_UTILITIES_INTN_T_HPP
+#define SHARE_UTILITIES_INTN_T_HPP
+
+#include "utilities/count_leading_zeros.hpp"
+
+#include <limits>
+
+template <unsigned int nbits>
+class uintn_t;
+
+// This class represents a signed integer type with the width of exactly nbits
+// bits. Conceptually, nbits == 8 gives a type equivalent to int8_t,
+// nbits == 16 gives a type equivalent to int16_t, and so on. This class may be
+// used to verify the correctness of an algorithm that is supposed to be
+// applicable to all fixed-width integral types. With a few bits, it makes it
+// possible to perform an exhaustive test that exercises the algorithm with all
+// possible input values.
+// Implementation-wise, this class currently only supports 0 < nbits <= 8. Also
+// note that this class is implemented so that overflows in alrithmetic
+// operations are well-defined and wrap-around.
```

Comment:
```suggestion
// operations are well-defined and wrap-around, just like jint, juint, jlong and julong.
```

Reply:
Overflow in `jint` and `jlong` is actually UB.

Reply:
Ah right. Only well defined with `java_add` etc.